internal/storage/image.go

```diff
@@ -365,44 +388,247 @@ func (svc *imageService) prepareReference(inputSystemContext *types.SystemContex
 }
 
 func (svc *imageService) PrepareImage(inputSystemContext *types.SystemContext, imageName string) (types.ImageCloser, error) {
-	systemContext, srcRef, err := svc.prepareReference(inputSystemContext, imageName)
+	systemContext, srcRef, err := svc.lookup.prepareReference(inputSystemContext, imageName)
 	if err != nil {
 		return nil, err
 	}
 
 	return srcRef.NewImage(svc.ctx, systemContext)
 }
 
-func (svc *imageService) PullImage(systemContext *types.SystemContext, imageName string, inputOptions *copy.Options) (types.ImageReference, error) {
-	policy, err := signature.DefaultPolicy(systemContext)
+// nolint: gochecknoinits
+func init() {
+	reexec.Register("crio-copy-image", copyImageChild)
+}
+
+type copyImageArgs struct {
+	Lookup        *imageLookupService
+	ImageName     string
+	ParentCgroup  string
+	SystemContext *types.SystemContext
+	Options       *ImageCopyOptions
+
+	StoreOptions storage.StoreOptions
+}
+
+// moveSelfToCgroup moves the current process to a new transient cgroup.
+func moveSelfToCgroup(cgroup string) error {
+	slice := "system.slice"
+	if rootless.IsRootless() {
+		slice = "user.slice"
+	}
+	if cgroup != "" {
+		if !strings.Contains(cgroup, ".slice") {
+			return fmt.Errorf("invalid systemd cgroup %q", cgroup)
+		}
+		slice = strings.ReplaceAll(strings.Trim(cgroup, "/"), "/", ":")
+	}
+
+	unitName := fmt.Sprintf("crio-pull-image-%d.scope", os.Getpid())
+
+	return utils.RunUnderSystemdScope(os.Getpid(), slice, unitName)
```

Comment:
We probably need the `CollectMode` set to `inactive-or-failed` for the pull so they don't leak on failures.

Reply:
Opened https://github.com/cri-o/cri-o/pull/4248 to support that.